packages/react-server/src/ReactFlightServer.js

```diff
@@ -110,6 +111,83 @@ import {SuspenseException, getSuspendedThenable} from './ReactFlightThenable';
 
 initAsyncDebugInfo();
 
+function patchConsole(consoleInst: typeof console, methodName: string) {
+  const descriptor = Object.getOwnPropertyDescriptor(consoleInst, methodName);
+  if (
+    descriptor &&
+    (descriptor.configurable || descriptor.writable) &&
+    typeof descriptor.value === 'function'
+  ) {
+    const originalMethod = descriptor.value;
+    const originalName = Object.getOwnPropertyDescriptor(
+      // $FlowFixMe[incompatible-call]: We should be able to get descriptors from any function.
+      originalMethod,
+      'name',
+    );
+    const wrapperMethod = function (this: typeof console) {
+      const request = resolveRequest();
+      if (methodName === 'assert' && arguments[0]) {
+        // assert doesn't emit anything unless first argument is falsy so we can skip it.
+      } else if (request !== null) {
+        // Extract the stack. Not all console logs print the full stack but they have at
+        // least the line it was called from. We could optimize transfer by keeping just
+        // one stack frame but keeping it simple for now and include all frames.
+        let stack = new Error().stack;
+        if (stack.startsWith('Error: \n')) {
+          stack = stack.slice(8);
+        }
+        const firstLine = stack.indexOf('\n');
+        if (firstLine === -1) {
+          stack = '';
+        } else {
+          // Skip the console wrapper itself.
+          stack = stack.slice(firstLine + 1);
+        }
+        request.pendingChunks++;
+        // We don't currently use this id for anything but we emit it so that we can later
+        // refer to previous logs in debug info to associate them with a component.
+        const id = request.nextChunkId++;
+        emitConsoleChunk(request, id, methodName, stack, arguments);
+      }
+      // $FlowFixMe[prop-missing]
+      return originalMethod.apply(this, arguments);
+    };
+    if (originalName) {
+      Object.defineProperty(
+        wrapperMethod,
+        // $FlowFixMe[cannot-write] yes it is
+        'name',
+        originalName,
+      );
+    }
+    Object.defineProperty(consoleInst, methodName, {
+      value: wrapperMethod,
+    });
+  }
+}
+
+if (
+  enableServerComponentLogs &&
+  __DEV__ &&
+  typeof console === 'object' &&
+  console !== null
+) {
+  // Instrument console to capture logs for replaying on the client.
+  patchConsole(console, 'assert');
+  patchConsole(console, 'debug');
+  patchConsole(console, 'dir');
+  patchConsole(console, 'dirxml');
+  patchConsole(console, 'error');
+  patchConsole(console, 'group');
+  patchConsole(console, 'groupCollapsed');
+  patchConsole(console, 'groupEnd');
+  patchConsole(console, 'info');
+  patchConsole(console, 'log');
+  patchConsole(console, 'table');
+  patchConsole(console, 'trace');
+  patchConsole(console, 'warn');
```

Comment:
There are some notable omissions from this list:

clear: This could make sense but it really refers to clearing the Flight runtime logs, it doesn't really imply that it should also clear the client.

time, timeLog, timeEnd, count, countReset: These run performance timings or state so it doesn't really make sense to replay these on the client. Instead we could run them in user space on the server and then write the result on the client.

profile, profileEnd, timeStamp: These really apply to profiles of an inspector running against the server so doesn't make sense to run on the client.

createTask and memory: These aren't side effects. You could log the result though. createTask might need deeper integration to forward the new stack.

context: This isn't really used anywhere. From what I can tell it's just built-in to v8 but without any UI. This could potentially actually be useful during replaying to put everything into an RSC context, depending on what the UI ends up looking like.